File: backend/agents/budget.py
"""Harness-enforced agent budget (rule 10: enforced in code, never by prompt).

Counts every tool call, sums cost points, and enforces a wall-clock deadline.
The clock is injectable so budgets can be unit-tested with no LLM and no real
time passing.
"""

from __future__ import annotations

import time
from typing import Callable


class BudgetExceeded(Exception):
    def __init__(self, reason: str, limit: float | int | None = None,
                 attempted: float | int | None = None) -> None:
        self.reason = reason
        self.limit = limit
        self.attempted = attempted
        super().__init__(f"budget exceeded: {reason} (limit={limit}, attempted={attempted})")


class Budget:
    def __init__(
        self,
        max_calls: int = 10,
        max_cost_points: int = 3,
        wall_clock_s: float = 60.0,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        self.max_calls = max_calls
        self.max_cost_points = max_cost_points
        self.wall_clock_s = wall_clock_s
        self._clock = clock
        self._start: float | None = None
        self.calls = 0
        self.cost = 0

    def start(self) -> "Budget":
        self._start = self._clock()
        return self

    def elapsed(self) -> float:
        return 0.0 if self._start is None else self._clock() - self._start

    def charge(self, cost_points: int = 0, label: str = "") -> "Budget":
        """Account for one tool call; raise BudgetExceeded if any limit trips.
        Nothing is incremented when a limit trips."""
        if self._start is None:
            self._start = self._clock()
        elapsed = self.elapsed()
        if elapsed > self.wall_clock_s:
            raise BudgetExceeded("wall_clock", self.wall_clock_s, round(elapsed, 3))
        if self.calls + 1 > self.max_calls:
            raise BudgetExceeded("max_calls", self.max_calls, self.calls + 1)
        if self.cost + cost_points > self.max_cost_points:
            raise BudgetExceeded("max_cost_points", self.max_cost_points, self.cost + cost_points)
        self.calls += 1
        self.cost += cost_points
        return self

    def wrap(self, fn: Callable, cost_points: int, label: str = "") -> Callable:
        def wrapped(*args, **kwargs):
            self.charge(cost_points, label=label or getattr(fn, "__name__", ""))
            return fn(*args, **kwargs)
        return wrapped

    def snapshot(self) -> dict:
        return {
            "calls": self.calls,
            "max_calls": self.max_calls,
            "cost": self.cost,
            "max_cost_points": self.max_cost_points,
            "elapsed_s": round(self.elapsed(), 3),
            "wall_clock_s": self.wall_clock_s,
        }

File: frontend/src/lib/budget.ts
/**
 * Agent budgets — MIRRORED BACKEND CONSTANTS, NOT SERVED DATA.
 *
 * Read this before trusting the meter.
 *
 * Rule 10 says budgets are enforced by the harness in code, and they are:
 * `Budget.snapshot()` (backend/agents/budget.py:68-76) returns exactly the six
 * numbers this meter wants — calls, max_calls, cost, max_cost_points, elapsed_s,
 * wall_clock_s. It has ZERO production callers. Nothing serialises it. No HTTP
 * response and no SSE payload carries a limit or a spend: `agent_step` is
 * {agent, tool, args_summary, result_summary} and `agent_done` is
 * {agent, status, summary}. `RunVerdict` does carry tool_calls and
 * cost_points_spent (investigator only), but GET /run/{id}/verdict returns just
 * {run_id, case_id, hypotheses, done} and drops them.
 *
 * So a live budget meter can only be reconstructed client-side: count the
 * agent_step events, map tool → cost with the table below, and compare against
 * limits copied out of the source. Every number here is a duplicate of a backend
 * constant and will silently rot if the backend changes.
 *
 * The worst offender: the investigator's max_cost_points is not a literal. It is
 * `autopilot_spend()` (investigator.py:87-89) = _CF_TOP_K * cost(run_counterfactual)
 * + cost(run_twin) = 5*1 + 2 = 7, computed at import time so that the agent and
 * the autopilot may spend exactly the same (see §The budget experiment in the
 * README). Retuning _CF_TOP_K moves it and this file will not notice.
 *
 * The honest fix is a backend one: emit budget.snapshot() on agent_step, or add
 * the spend + limits to the verdict endpoint. Until then the meter is labelled
 * as mirrored in the UI.
 */
import type { AgentName } from '@/types/api'

/**
 * backend/agents/tools.py:381-395. Only three tools cost anything; every other
 * registered tool is 0.
 */
export const TOOL_COST: Readonly<Record<string, number>> = {
  run_twin: 2,
  run_counterfactual: 1,
  rehearse_fix: 1,
}

export function costOf(tool: string | null | undefined): number {
  if (!tool) return 0
  return TOOL_COST[tool] ?? 0
}

export function isExpensive(tool: string | null | undefined): boolean {
  return costOf(tool) > 0
}

export interface AgentBudget {
  maxCalls: number
  maxCostPoints: number
  wallClockS: number
  /** Where the numbers come from, shown in the UI so nobody trusts them blindly. */
  source: string
}

/**
 * Per-agent limits, copied from source. The narrator is absent on purpose: it
 * is never registered in RunVerdict.transcripts, so its transcript endpoint
 * always 404s.
 */
export const AGENT_BUDGET: Readonly<Record<AgentName, AgentBudget>> = {
  investigator: {
    maxCalls: 16,
    maxCostPoints: 7,
    wallClockS: 180,
    source: 'investigator.py:92-113 — max_cost_points = autopilot_spend() = 5×1 + 2',
  },
  challenger: {
    maxCalls: 5,
    maxCostPoints: 0,
    wallClockS: 30,
    source: 'challenger.py:94 — the challenger may spend nothing at all',
  },
  remediation: {
    maxCalls: 6,
    maxCostPoints: 3,
    wallClockS: 45,
    source: 'remediation.py:168',
  },
}

export interface AgentSpend {
  calls: number
  points: number
}

/** Reconstructed from the agent_step stream — the only live signal there is. */
export function spendFor(steps: { tool: string | null }[]): AgentSpend {
  return {
    calls: steps.length,
    points: steps.reduce((total, s) => total + costOf(s.tool), 0),
  }
}

/**
 * A `budget_exhausted` agent_done carries the real limit, but only as prose in
 * `summary`: "budget exceeded: {reason} (limit={limit}, attempted={attempted})"
 * (budget.py:20). It is the one place a limit reaches the client, and only when
 * the budget trips. Parsed leniently: a miss just means no extra detail.
 */
export interface BudgetTrip {
  reason: string
  limit: string
}

export function parseBudgetTrip(summary: string | null): BudgetTrip | null {
  if (!summary) return null
  const m = /budget exceeded:\s*(\w+)\s*\(limit=([^,)]+)/.exec(summary)
  if (!m) return null
  return { reason: m[1]!, limit: m[2]! }
}
